SCSS: resolve bare module imports on Windows (#245579)

IsaacTian05 opened the pull request:
Fix issue https://github.com/microsoft/vscode-css-languageservice/issues/432

IsaacTian05 committed:
fix(scss): resolve bare-module imports to node_modules on all platforms (#245579)

IsaacTian05 committed:
test fix(scss): resolve bare-module imports to node_modules on all platforms (#245579)

IsaacTian05 committed:
Create .travis.yml

IsaacTian05 committed:
Create triggerbuild.txt

IsaacTian05 committed:
Create .travis.yml

aeschli commented on src/test/scss/scssNavigation-node-modules.test.ts:
```diff
@@ -0,0 +1,17 @@
+    const doc = TextDocument.create('file:///c:/proj/app.scss', 'scss', 1,
+        "@import 'bootstrap/scss/variables';");
+    const links = await ls.findDocumentLinks2(doc, ls.parseStylesheet(doc), {}, mockFS);
+    const target = links[0].target!.replace(/\\/g, '/');
```
target is a URI, it should not contain `\`

IsaacTian05 committed:
Delete .travis.yml

IsaacTian05 committed:
Delete .travis.yml

IsaacTian05 committed:
Change target (URI) to not contain '\' character

aeschli commented on triggerbuild.txt:
```diff
@@ -0,0 +1 @@
+build
```
This should also be removed....

aeschli commented on src/services/cssNavigation.ts:
```diff
@@ -422,7 +422,20 @@ export class CSSNavigation {
+		// Treat bare module names (“bootstrap/...”) like sass-loader does
+		const isBareImport = !target.startsWith('.')             // not ./ or ../
+		                 && !target.startsWith('/')              // not workspace-absolute
+		                 && target.indexOf(':') === -1;          // not a scheme (file://)
```
Maybe a use a regex to make that check [
startsWithSchemeRegex = /^\w[\w\d+.-]:/

aeschli reviewed (changes requested):
Looks good. If you can add some more test cases that would be great

IsaacTian05 committed:
Delete triggerbuild.txt

IsaacTian05 committed:
fix cssNavigation: detect scheme‑prefixed import paths with regex

IsaacTian05 committed:
test scss‑navigation : verify scheme‑prefixed @import paths remain absolute

Adds `schemeImportLinks.test.ts` with four mocha cases covering
`http://`, `https://`, `file://`, and `vscode-resource://` URLs

aeschli commented on src/services/cssNavigation.ts:
```diff
@@ -422,7 +422,22 @@ export class CSSNavigation {
 
-		const ref = await this.mapReference(documentContext.resolveReference(target, documentUri), isRawLink);
+		// Treat bare module names (“bootstrap/...”) like sass-loader does
+		const startsWithSchemeRegex = /^\w[\w\d+.-]:/;
```
Node module resolving should only happen when `this.resolveModuleReferences` is true.
(only set for LESS and SCSS).

So this needs to go behind the `if (this.resolveModuleReferences) {`.
I looks like we already call `resolveModuleReference` there, so this code is not necessary?

IsaacTian05 reviewed (commented):
Pending

miss-programgamer commented:
If these changes are still pending, I can go ahead and implement them.

...Actually, how *would* I go about adding changes to someone else's pull request?

miss-programgamer committed:
Implemented changes requested in pr #440

miss-programgamer committed:
Merge branch 'main' of github.com:miss-programgamer/vscode-css-languageservice into fix/scss-import-node-modules

miss-programgamer committed:
Merge branch 'main' of github.com:microsoft/vscode-css-languageservice into fix/scss-import-node-modules

miss-programgamer committed:
Merge branch 'main' into fix/scss-import-node-modules

IsaacTian05 committed:
Merge pull request #1 from miss-programgamer/fix/scss-import-node-modules

Implemented changes requested in pr #440
